pkg/api/steve/projectresources/handlers.go

```diff
@@ -0,0 +1,853 @@
+				Name:      service.Name,
+			},
+		},
+	}
+	if tlsSecret == nil { // in dev mode, there is no CA
+		apiService.Spec.InsecureSkipTLSVerify = true
+	} else {
+		apiService.Spec.CABundle = tlsSecret.Data[corev1.TLSCertKey]
+	}
+	// special non-project namespace to "contain" namespaces not in a project
+	namespace := corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: unscopedNamespace,
+			Annotations: map[string]string{
+				NamespaceAnnotation: "true",
+			},
+		},
+	}
+	// cluster role for listing resources in this APIGroup
+	role := rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: RoleName,
+			Annotations: map[string]string{
+				AuthzAnnotation: "unscoped",
+			},
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				Verbs:     []string{"list"},
+				APIGroups: []string{Group},
+				Resources: []string{"*"},
+			},
+		},
+	}
+	// default rolebinding for the special unscoped namespace, the prtb controller will create similar rolebindings for project namespaces for specific users
+	roleBinding := rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      RoleName,
+			Namespace: unscopedNamespace,
+			Annotations: map[string]string{
+				AuthzAnnotation: "unscoped",
+			},
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind: "Group",
+				Name: "system:authenticated", // all users can access this endpoint initially, the handler will do a SAR check before returning resources
+			},
+		},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: rbacv1.GroupName,
+			Kind:     "ClusterRole",
+			Name:     RoleName,
+		},
+	}
+	err = a.apply.ApplyObjects(&apiService, &namespace, &role, &roleBinding)
+	return service, err
+}
+
+func getDynamicClient(config *wrangler.Context) (dynamic.Interface, error) {
+	restConfig, err := config.ClientConfig.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+	return dynamic.NewForConfig(restConfig)
+}
+
+// authenticateAPIServer is a middleware that verifies the client's certificates and checks that the client's CN is in the API server's allow list.
+// This fulfills the two conditions outlines in https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer/#kubernetes-apiserver-client-authentication:
+// > 1. The connection must be made using a client certificate that is signed by the CA whose certificate is in --requestheader-client-ca-file.
+// > 2. The connection must be made using a client certificate whose CN is one of those listed in --requestheader-allowed-names.
+// The http.Server must have a TLS config with ClientAuth: tls.RequestClientCert set in order for the http.Request to have the client cert populated.
+// Normally mTLS would be handled at the network session layer automatically during the client connection when tls.RequireAndVerifyClientCert
+// is set on the Server's tls.Config, but since this server performs multiple functions and not all of them require mTLS, we handle it in the
+// application layer in this middleware.
```

Comment:
Here's an alternative to this client tls verification: https://github.com/cmurphy/rancher/commit/891d0caf5e6416dc1441c56db633fc920f72eee9

Benefits of the alternative:
- relies on the stdlib to do client cert verification, less likely to introduce security holes
Drawbacks:
- loads the CA only on startup, if the CA is changed in the configmap it won't be used until rancher reloads
- it spreads out the logic to two places, here and pkg/tls, when it really only applies to this one endpoint. Most of rancher's tls has zero to do with the k8s API server's CA, so it's out of place to add that verification there.